docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+
+1. Provide a provably correct inference algorithm for Flux.
+2. Represent a table stream as a list of record types and update the type signatures for each Flux transformation accordingly.
+
+As outlined in [#1578](https://github.com/influxdata/flux/issues/1578), in order to be able to prove the correctness of type inference, we must break the mutual recursion between _type_ unification and _kind_ unification.
+And to do that, we must replace our current record _kind_ with an extensible record _type_.
+
+### Extensible Records
+
+The Flux type system and the current Flux record _kind_ is based on the following [paper](https://caml.inria.fr/pub/papers/garrigue-structural_poly-fool02.pdf).
+In this system record kinds have a set of properties along with their types.
+They also contain two sets of labels.
+One set specifies the properties that a record **must** have, the other specifies the properties it **may** have.
+While this allows for polymorphic record access as well as polymorphic record literals, it lacks support for record extensions.
+The following example cannot be typed in this system.
+
+```
+r = {a: 1, b: 2}
+s = {r with c: 3} // s is a record with properties a, b, and c
+```
+
+In order to support examples like the one above, I propose we replace the current record kind with one that allows for extension such as the one in [this paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/scopedlabels.pdf).
+Here the syntax of record types more closely aligns with the syntax of the values themselves.
+In this system records are constructed via extension, starting from the empty record `{}`.
+
+For example the record literal `{a:1, b:2}` is just shorthand for `{a:1 | {b:2 | {} }}` whose type is given by `{a:int | {b:int | {} }}`.
+
+More generally, the polymorphic type of the extension operation `{_ with l:_}` is given by:
+
+```
+forall ['a, 'r] where 'r: Rec ('a, 'r) -> {l: 'a | 'r}
+```
+
+This type is read as, "for all type variables _'a_ and _'r_ where _'r_ is a record kind, we have a function type taking _'a_ and _'r_ as input and returning a record type extended from _'r_ by adding a field _l_ of type _'a_".
+Selection and restriction operations are defined similarly, and update and rename can be expressed using these basic operations.
+
+Unification of record types is particularly elegant and can be expressed using the following 3 rules:
+
+1. `{a: 't | 'r} = {a: 'u | 's} => 't = 'u and 'r = 's`
+2. `{a: 't | 'r} = {b: 'u | 's} => {a: 't | 'v} = 's and {b: 'u | 'v} = 'r`
+3. `{a: 't | 'r} = {b: 'u | 'r} => error`
+
+Note here `=` is synonymous for unify.
+
+### Implications
+
+The Flux type system extends Hindley-Milner with a _kind_ system.
+Currently records are modeled as _kinds_ in this system.
+Extensible records however fall entirely within the _type_ system, not the _kind_ system.
+Describing records this way allows us to break the mutual recursion between type unification and kind unification.
```

Comment:
What about the function kind? It does a similar thing were it models its parameters as  a lower/upper bound of set of parameters. Does the mutual recursion not occur because parameter type vars are never free?

Reply:
In the current system functions are actually _types_ not _kinds_. Hence there's no mutual recursion introduced and we can keep the current definition and not lose parity.